lls/cache.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_LLS_CACHE_H_
+#define _GATEKEEPER_LLS_CACHE_H_
+
+#include "gatekeeper_lls.h"
+#include "gatekeeper_net.h"
+
+/*
+ * Maximum key length for an LLS cache entry, used to determine
+ * how much memory should be allocated for cache requests.
+ * Therefore, it should be set to the maximum key length
+ * needed, but no more to avoid wasting mempool allocation.
+ */
+#define LLS_CACHE_MAX_KEY_LEN	4
+
+enum lls_cache_op {
+	/* Add an entry to the cache. */
+	LLS_CACHE_ADD,
+	/* Increment the refcnt of a cache entry. */
+	LLS_CACHE_INC,
+	/* Decrement the refcnt of a cache entry. */
+	LLS_CACHE_DEC,
+};
+
+struct lls_cache_req {
+	/* Entry to be stored. */
+	struct lls_cache_entry *entry;
+
+	/* Operation of this request. */
+	enum lls_cache_op      op;
+
+	/* The cache that this request will change. */
+	struct lls_cache       *cache;
+
+	/*
+	 * IP address for a new entry (or key for the hash).
+	 *
+	 * NOTE
+	 *	This is a variable-length field.
+	 */
+	const void             *ip_key[0];
```

Comment:
Given that you have `LLS_CACHE_MAX_KEY_LEN`, do you still need this variable length field?